pykg2tbl/kg2tbl.py

```diff
@@ -0,0 +1,233 @@
+import logging
+from abc import ABC, abstractmethod
+from typing import Callable, Generator, Iterable, List, Tuple, Union
+
+from rdflib import Graph
+from SPARQLWrapper import SPARQLWrapper
+
+from pykg2tbl.exceptions import (
+    CompatibilityCheckerNotCallable,
+    MultipleSourceTypes,
+    NoCompatibilityChecker,
+    NotASubClass,
+    WrongInputFormat,
+)
+from pykg2tbl.query import QueryResult
+
+log = logging.getLogger(__name__)
+
+
+# Create abstract class for making a contract by design for devs ##
+class KGSource(ABC):
+    @abstractmethod
+    def query_result_to_list_dicts(reslist: list) -> List:
+        """
+        From the query result build a standard list of dicts,
+            where each key is the relation in the triplet.
+
+        :param reslist: list with the query.
+        """
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def query(self, sparql: str) -> QueryResult:
+        """
+        Function that queries data with the given sparql
+
+        :param sparql: sparql statement logic for querying data.
+        """
+        pass  # pragma: no cover
+
+    def exec(self, query: str, output_file: str, sep: str):
+        """
+        Service that will make query a provided kgsource and
+            export a tabular data file based on the users preferences.
+
+        :param str query: named template sparql
+        :param str output_file: file to write query output as a table
+        :param str sep: table separator
+
+        """
+        result = self.query(query)
+        result.as_csv(output_file, sep)
+
+    registry = set()
+
+    @staticmethod
+    def register(constructor):
+        # assert that constructor is for a subclass of QueryResult
+        # e.g. check if method check_compatibility is present
+        if not issubclass(constructor, KGSource):
+            raise NotASubClass(parent_class="KGSource")
+        check_compatibility_function = getattr(
+            constructor, "check_compatibility", False
+        )
+        if not check_compatibility_function:
+            raise NoCompatibilityChecker
+        if not isinstance(check_compatibility_function, Callable):
+            raise CompatibilityCheckerNotCallable
+        KGSource.registry.add(constructor)
+
+    @staticmethod
+    def build(*files):
```

Comment:
`files` ?
doesn´t feel like the correct name for these "strings that represent sources"
suggesting `sources` or `src_str` if you fear people expect the former to represent KGSource instances